### PR TITLE
Const Compare for Tuples

### PR DESCRIPTION
**Pull request by onestacked**
Makes the impls for Tuples of ~const `PartialEq` types also `PartialEq`, impls for Tuples of ~const `PartialOrd` types also `PartialOrd`, for Tuples of ~const `Ord` types also `Ord`.

behind the `#![feature(const_cmp)]` gate.

~~Do not merge before #104113 is merged because I want to use this feature to clean up the new test that I added there.~~

r? @fee1-dead

### CODE REVIEW

File: src/test/ui/consts/fn_trait_refs.rs [hunk 1]
```diff
@@ -68,13 +68,9 @@ const fn main() {
...
     let test_one = test_fn(one);
-    assert!(test_one.0 == 1);
-    assert!(test_one.1 == 1);
-    assert!(test_one.2 == 1);
+    assert!(test_one == (1, 1, 1));
 
     let test_two = test_fn_mut(two);
-    assert!(test_two.0 == 1);
-    assert!(test_two.1 == 1);
+    assert!(test_two == (2, 2));
```
Comment (fee1-dead) [inline src/test/ui/consts/fn_trait_refs.rs:75]: These asserts don't work unless you put them in a const item, `const _: () = {}`

Comment (chriss0612) [reply src/test/ui/consts/fn_trait_refs.rs:75]: They are in const fn main()

Comment (fee1-dead) [reply src/test/ui/consts/fn_trait_refs.rs:75]: they don't work that way

Comment (chriss0612) [reply src/test/ui/consts/fn_trait_refs.rs:75]: ahh right that only checks that i builds

**Pull request merged**